### PR TITLE
Reduce type guessing in remote

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Reduce use of guessing Python types from Flyte types in the Flyte remote experience.  Also some minor cleanup to the code.

## Type
 - [x] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

* Added some fake integration tests in the unit test folder that are marked as `sandbox_test`s.  Updated the Makefile to skip these.
* Update the `LiteralsResolver` object to support caching the native values.  Each Literal only gets converted once.  Guessing is still used but it's a fallback.
* Considered making the `LiteralsResolver` inherit from `dict` directly, but doesn't seem to buy us anything and just adds complexity.
* Update the interfaces for the remote execution objects that have inputs/outputs to return the `LiteralsResolver` object directly.
* Move common execution functionality into a base class `RemoteExecutionBase`.
* Remove singledispatch from execute and just call each function directly.  Also update the names to make more sense.
* In `TypeEngine.dict_to_literal_map()` change the `guessed_python_types` argument name to `type_map` to match other usage.  Updated tests to match... hoping that most people are not using this argument as a kwarg and just a regular arg.  Hopefully most people aren't even using this function.
* Removing the `guessed_python_interface` properties in `FlyteLaunchPlan`, `FlyteWorkflow` and `FlyteTask`.
* In `register_workflow`, call `create_workflow` only after downstream entities are registered.
* In `register_workflow` update `all_downstream` to True by default.
* Removing `from_project` and `from_domain` in a bunch of places.  It seemed like it wasn't really used in most cases?

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2208

### CODE REVIEW

File: flytekit/remote/executions.py [hunk 1]
```diff
@@ -1,58 +1,120 @@
...
-        """
-        Returns the outputs of the task execution, if available, in the standard Python format that is produced by
-        the type engine.
+    @abstractmethod
+    def error(self) -> core_execution_models.ExecutionError:
+        ...
+
+    @property
+    @abstractmethod
+    def is_complete(self) -> bool:
```
Comment (kumare3) [inline flytekit/remote/executions.py:66]: is_done()

File: flytekit/remote/remote.py [hunk 2]
```diff
@@ -441,43 +433,42 @@ def register_workflow(
...
+
         return self.fetch_workflow(ident.project, ident.domain, ident.name, ident.version)
 
     def register_launch_plan(
         self,
         entity: LaunchPlan,
-        serialization_settings: typing.Optional[SerializationSettings] = None,
-        version: typing.Optional[str] = None,
+        version: str,
+        project: typing.Optional[str] = None,
```
Comment (kumare3) [inline flytekit/remote/remote.py:455]: maybe have the serialization_settings so that we can register the workflow and tasks

File: flytekit/remote/remote.py [hunk 3]
```diff
@@ -680,39 +694,95 @@ def execute(
...
         project: str = None,
         domain: str = None,
-        name: str = None,
-        version: str = None,
         execution_name: str = None,
-        from_project: str = None,
-        from_domain: str = None,
         options: typing.Optional[Options] = None,
         wait: bool = False,
+        type_map: typing.Optional[typing.Dict[str, typing.Type]] = None,
```
Comment (kumare3) [inline flytekit/remote/remote.py:780]: type hints

File: flytekit/remote/remote.py [hunk 4]
```diff
@@ -1206,25 +1341,16 @@ def _assign_inputs_and_outputs(
         self,
         execution: typing.Union[FlyteWorkflowExecution, FlyteNodeExecution, FlyteTaskExecution],
         execution_data,
-        interface,
+        interface: TypedInterface,
     ):
         """Helper for assigning synced inputs and outputs to an execution object."""
-        with self.remote_context() as ctx:
+        with self.remote_context():
```
Comment (kumare3) [inline flytekit/remote/remote.py:1347]: yuck!

File: flytekit/remote/remote.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/remote/remote.py:455]: can i ignore this and leave as is since this will touch your pr anyways?

File: flytekit/remote/remote.py [hunk 5]
```diff
@@ -621,22 +618,27 @@ def execute(
         name: str = None,
         version: str = None,
         execution_name: str = None,
-        from_project: str = None,
-        from_domain: str = None,
+        image_config: typing.Optional[ImageConfig] = None,
```
Comment (wild-endeavor) [inline flytekit/remote/remote.py:621]: I'm adding this here @kumare3 and @eapolinario

File: flytekit/core/type_engine.py [hunk 6]
```diff
@@ -1428,22 +1429,100 @@ class LiteralsResolver(object):
     LiteralsResolver is a helper class meant primarily for use with the FlyteRemote experience or any other situation
     where you might be working with LiteralMaps. This object allows the caller to specify the Python type that should
     correspond to an element of the map.
-    TODO: Add an optional Flyte idl interface model object to the constructor
+
+    TODO: Consider inheriting from collections.UserDict instead of manually having the _native_values cache
     """
 
-    def __init__(self, literals: typing.Dict[str, Literal]):
+    def __init__(
```
Comment (kumare3) [inline flytekit/core/type_engine.py:1436]: docs on what variable map is?

Comment (wild-endeavor) [reply flytekit/core/type_engine.py:1436]: can i add in a future pr?

Comment (wild-endeavor) [reply flytekit/core/type_engine.py:1436]: nm added a quick comment

**Review by eapolinario — APPROVED** [review]
